src/Common/src/TypeSystem/Common/VirtualMethodAlgorithm.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Pluggable virtual method computation algorithm.
+    /// </summary>
+    public abstract class VirtualMethodAlgorithm
+    {
+        /// <summary>
+        /// Resolves interface method '<paramref name="interfaceMethod"/>' to a method on '<paramref name="type"/>'
+        /// that implements the the method.
+        /// </summary>
+        public abstract MethodDesc ResolveInterfaceMethodToVirtualMethodOnType(MethodDesc interfaceMethod, TypeDesc currentType);
+
+        /// <summary>
+        /// Resolves a virtual method call.
+        /// </summary>
+        public abstract MethodDesc FindVirtualFunctionTargetMethodOnObjectType(MethodDesc targetMethod, TypeDesc objectType);
+
+        /// <summary>
```

Comment:
I don't believe ComputeAllVirtualMethods(TypeDesc type) should be placed on this algorithm. I believe it should be in its own algorithm ... perhaps VirtualMethodsEnumerationAlgorithm or something such.